tree/ntupledraw/v7/src/RDrawStorage.cxx

```diff
@@ -0,0 +1,601 @@
+   fTH1Fs.back()->SetStats(0);
+   fTH1Fs.back()->DrawCopy();
+
+   // 5. Draw all boxes and add possibility to click on RPageBox to obtain information about a page
+   fHeaderBox->Draw();
+   for (const auto &b : fPageBoxes) {
+      b->Draw();
+   }
+   fFooterBox->Draw();
+   fPad->AddExec("ShowPageDetails", "ROOT::Experimental::Detail::RDrawStorage::RPageBoxClicked()");
+
+   // 6. Draw clusterAxis
+   // fTexts.at(0) points to Title so skip
+   for (std::size_t i = 1; i < fTexts.size(); ++i) {
+      fTexts.at(i)->Draw();
+   }
+   for (const auto &l : fLines) {
+      l->Draw();
+   }
+   fPad->Update();
+
+   // 7. Return to canvas, draw title, legend and description of x-axis
+   fCanvasPtrs.back()->cd();
+   fTexts.at(0)->Draw(); // Title
+   fLegend->Draw();
+   TLatex latex;
+   latex.SetTextSize(0.04);
+   latex.DrawLatex(0.955, 0.5, fAxisTitle.c_str());
+
+   fCanvasPtrs.back()->Update();
+}
+
+void ROOT::Experimental::Detail::RDrawStorage::SetPageIds()
+{
+   for (std::size_t i = 0; i < fPageBoxes.size(); ++i) {
+      fPageBoxes.at(i)->SetPageId(i + 1);
+   }
+}
+
+void ROOT::Experimental::Detail::RDrawStorage::RPageBoxClicked()
+{
+   int event = gPad->GetEvent();
+   if (event != kButton1Up)
+      return;
+   TObject *select = gPad->GetSelected();
+   if (!select)
+      return;
+   if (select->InheritsFrom(ROOT::Experimental::Detail::RPageBox::Class())) {
+      ROOT::Experimental::Detail::RPageBox *pageBox = (ROOT::Experimental::Detail::RPageBox *)select;
+      pageBox->Inspect();
+   } else if (select->InheritsFrom(ROOT::Experimental::Detail::RMetaDataBox::Class())) {
+      ROOT::Experimental::Detail::RMetaDataBox *metaBox = (ROOT::Experimental::Detail::RMetaDataBox *)select;
+      metaBox->Inspect();
+   }
+}
+
+std::int32_t ROOT::Experimental::Detail::RDrawStorage::GetColourFromFieldId(ROOT::Experimental::DescriptorId_t fieldId)
+{
+   std::int32_t colour = 0;
+   fieldId %= 61;
+   switch (fieldId % 12) {
+   case 0: colour = kRed; break;
+   case 1: colour = kMagenta; break;
+   case 2: colour = kBlue; break;
+   case 3: colour = kCyan; break;
+   case 4: colour = kGreen; break;
+   case 5: colour = kYellow; break;
+   case 6: colour = kPink; break;
+   case 7: colour = kViolet; break;
+   case 8: colour = kAzure; break;
+   case 9: colour = kTeal; break;
+   case 10: colour = kSpring; break;
+   case 11: colour = kOrange; break;
+   default:
+      // never here
+      assert(false);
+      break;
+   }
+   switch (fieldId / 12) {
+   case 0: colour -= 2; break;
+   case 1: break;
+   case 2: colour += 3; break;
+   case 3: colour -= 6; break;
+   case 4: colour -= 9; break;
+   case 5:
+      if (fieldId == 60)
+         return kGray;
+   default:
+      // never here
+      assert(false);
+      break;
+   }
+   return colour;
+}
+
+// ------------------------------ RNTupleDraw -------------------------------
+
+ROOT::Experimental::RNTupleDraw::RNTupleDraw(const std::unique_ptr<RNTupleReader> &reader)
```

Comment:
Can we just pass an `RNTupleReader *reader`?  The Draw object does not own the reader, correct?

Reply:
Originally I also wanted to just pass a raw pointer to avoid complications. But since `RNTupleReader::Open()` returns a unique_ptr and there might be non-C++ inclined users who don't know how to get the raw pointer form a unique_ptr, I thought it would be less confusing for users this way. I would leave the decision up to you if it should accept a raw pointer, reference to unique_ptr or allow both.